api/src/opentrons/protocol_engine/state/_labware_origin_math.py

```diff
@@ -0,0 +1,236 @@
+
+from opentrons.types import Point
+from opentrons_shared_data.labware.labware_definition import (
+    LabwareDefinition,
+    LabwareDefinition2,
+    LabwareDefinition3,
+    Vector3D,
+)
+from opentrons_shared_data.deck.types import DeckDefinitionV5
+from ..types import (
+    LabwareParentDefinition,
+    ModuleDefinition,
+    ModuleModel,
+    LabwareOffsetVector,
+    DeckLocationDefinition,
+    LabwareLocation,
+    ModuleLocation,
+    DeckSlotLocation,
+    AddressableAreaLocation,
+    OnLabwareLocation,
+)
+
+
+@overload
+def get_parent_placement_origin_to_lw_origin(
+    child_labware: LabwareDefinition,
+    parent_entity: ModuleDefinition,
+    module_parent_to_child_offset: LabwareOffsetVector,
+    deck_definition: DeckDefinitionV5,
+    is_topmost_labware: bool,
+    labware_location: ModuleLocation,
+) -> Point:
+    ...
+
+
+@overload
+def get_parent_placement_origin_to_lw_origin(
+    child_labware: LabwareDefinition,
+    parent_entity: DeckLocationDefinition,
+    module_parent_to_child_offset: None,
+    deck_definition: DeckDefinitionV5,
+    is_topmost_labware: bool,
+    labware_location: Union[DeckSlotLocation, AddressableAreaLocation],
+) -> Point:
+    ...
+
+
+@overload
+def get_parent_placement_origin_to_lw_origin(
+    child_labware: LabwareDefinition,
+    parent_entity: LabwareDefinition,
+    module_parent_to_child_offset: None,
+    deck_definition: DeckDefinitionV5,
+    is_topmost_labware: bool,
+    labware_location: OnLabwareLocation,
+) -> Point:
+    ...
+
+
+def get_parent_placement_origin_to_lw_origin(
+    child_labware: LabwareDefinition,
+    parent_entity: LabwareParentDefinition,
+    module_parent_to_child_offset: Union[LabwareOffsetVector, None],
+    deck_definition: DeckDefinitionV5,
+    is_topmost_labware: bool,
+    labware_location: LabwareLocation,
+) -> Point:
+    """Returns the offset from parent entity's placement origin to child labware origin.
+
+    Placement origin varies depending on the parent entity type (labware v3 are the back left bottom, and
+    labware v2, modules, & deck location types are the front left bottom).
+
+    Only parent-child specific offsets are calculated. Offsets that apply to a single entity
+    (ex., module cal) or the entire stackup (ex., LPC) are handled elsewhere.
+    """
+    parent_entity_origin_to_child_labware_placement_origin = (
+        _get_parent_entity_origin_to_child_labware_placement_origin(
+            child_labware=child_labware,
+            parent_entity=parent_entity,
+            module_parent_to_child_offset=module_parent_to_child_offset,
+            deck_definition=deck_definition,
+            labware_location=labware_location,
+        )
+    )
+
+    if isinstance(child_labware, LabwareDefinition2):
+        # For v2 definitions, cornerOffsetFromSlot is the parent entity placement origin to child labware origin offset.
+        # We only consider it when the child labware is the topmost labware in a stackup.
```

Comment:
Whew, thanks for identifying this caveat about `cornerOffsetFromSlot` only applying to the topmost labware.

Small suggestion to make it absolutely clear that we're not doing this "intentionally":

```suggestion
        # For v2 definitions, cornerOffsetFromSlot is the parent entity placement origin to child labware origin offset.
        # For compatibility with historical (buggy?) behavior,
        # we only consider it when the child labware is the topmost labware in a stackup.
```